core/src/main/scala/kafka/log/Log.scala

```diff
@@ -357,12 +357,47 @@ class Log(@volatile var dir: File,
 
   def lastStableOffsetLag: Long = highWatermark - lastStableOffset
 
-  def offsetSnapshot: LogOffsetSnapshot = {
+  def fullOffsetSnapshot: LogOffsetSnapshot = {
+    val hw: LogOffsetMetadata = if (_highWatermarkMetadata.messageOffsetOnly) {
+      val fullOffset = convertToOffsetMetadataOrThrow(_highWatermarkMetadata.messageOffset,
+        "Could not load offset metadata for high watermark")
+      lock.synchronized {
+        _highWatermarkMetadata = fullOffset
```

Comment:
`maybeFetchHighWatermarkOffsetMetadata` is used when we initialized a replica as a leader. It loads the HW metadata and falls back to LSO if it's out of range. Seems like enough of a separate usage to keep separate maybe?

But now that you mention it, should `maybeFetchHighWatermarkOffsetMetadata` be synchronized as well? It probably only gets called by a single thread (Partition#makeLeader), but that doesn't guard against future usages.

Reply:
I guess it depends on whether the fallback logic makes sense in this context. I guess it is reasonable to raise an out of range error here since we expect the high watermark to exist in the log. 

Agreed that `maybeFetchHighWatermarkOffsetMetadata` should be locked. Possibly we skipped it because this was only used in initialization. I don't see the harm adding it now.

Reply:
Were we going to add the locking to `initializeHighWatermarkOffsetMetadata`?